src/Tinfoil/Data/MAC.hs

```diff
@@ -0,0 +1,48 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DeriveGeneric #-}
+module Tinfoil.Data.MAC(
+    SigningKey(..)
+  , MAC(..)
+  , KeyedHashFunction(..)
+  , parseKeyedHashFunction
+  , renderKeyedHashFunction
+  ) where
+
+import           Control.DeepSeq.Generics (genericRnf)
+
+import           Data.ByteString (ByteString)
+
+import           GHC.Generics (Generic)
+
+import           P
+
+newtype SigningKey =
```

Comment:
I think it would be worth pulling out a Key module with different key types, and work on polishing terminology a but, I am not sure that "signing" is a good discriminator, at least as it is used at the moment for both symmetric and asymmetric use. There is a lack of consistent terminology for symmetric keys in general, commonly SecretKey or SymmetricKey is used, I think I lean towards SymmetricKey, I think it is useful to differentiate in types because the key generation process is so different.